ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+  def some_signal(arg)
+    raise NoMethodError
+  end
+
+  workflow_query 'custom-query-name'
+  def some_query(arg)
+    raise NoMethodError
+  end
+end
+```
+
+And client usage may look like:
+
+```ruby
+# Connect client
+client = Client.connect('localhost:7233')
+
+# Start workflow
+handle = client.start_workflow(
+  MyWorkflow,
+  "some arg",
+  id="wf-#{SecureRandom.uuid}",
+  task_queue='my_task_queue',
+)
+
+# Send signal
+handle.signal(MyWorkflow.some_signal, "some arg")
+```
+
+* This leverages a `method_added` approach to "decorate" methods.
+* This will probably create class-method equivalents for each item that will return "definitions" for use in clients.
+* We will not spend much time getting this right at first.
+
+## Data Conversion
+
+Like other SDKs, the following classes will exist:
+
+* `Temporalio::Converter::DataConverter` - Accepts a payload converter (must be Ractor shareable), failure converter
+  (must be Ractor shareable), and a payload codec
+  * 💭 Why use `Temporalio::Converter` module instead of just `Temporalio::DataConverter`? There are many files that are
+    included in conversion and most regular users won't ever use them.
+  * 💭 Why duplicate words instead of `Temporalio::Converter::Data`? `Temporalio::Converter` is not a base class and a
+    `DataConverter` is a completely separate thing from, say, a `PayloadConverter`.
+  * ❓ Would we prefer `Temporalio::Conversion` as the module name?
+* `Temporalio::Converter::PayloadConverter` - Default composite payload implementation accepting a collection of
+  encoding payload converters accepted in constructor, with a default class attr that has a default set.
+  * Users are expected to extend this and override conversion methods as needed. Default instantiation of this is 0
+    encoding converters and a `NotImplementedError` if either is called.
+  * 💭 Why not have a separate base class instead of combining the composite one with the one to override? This was a
+    50/50 call, but most users will probably add an encoding impl instead of override at this level. This is admittedly
+    a slight deviation from other SDKs.
+* `Temporalio::Converter::PayloadConverter::Encoding` - Base unimplemented encoding converter
+  * ❓ Would we prefer `Temporalio::Converter::EncodingPayloadConverter`?
+* `Temporalio::Converter::PayloadConverter::JSON`, `Temporalio::Converter::PayloadConverter::JSONProto`, etc  - All
+  common encoding converters
+* `Temporalio::Converter::FailureConverter` - Default failure converter, can be overridden
+* `Temporalio::Converter::PayloadCodec` - Unimplemented base codec
+  * Important for implementers to ensure this is async capable if using async activities anywhere
+* `Temporalio::Converter::RawValue` - Can be used as way to move around unconverted payloads
+
+#### Temporalio::Converter::PayloadConverter::JSON
+
+* Ruby has a `JSON` module in the standard library, but by default it's only primitives, arrays, hashes, etc and not
+  classes.
```

Comment:
What about JSON Protobuf, I'm assuming not supported...

Also, what are the implications for cross-language integration. For example, an activity in python that returns a json serialized object called from a Ruby workflow. Is the workaround to manually create the ruby object from json?

Reply:
> What about JSON Protobuf, I'm assuming not supported...
JSON protobuf will be supported out of the box

> Also, what are the implications for cross-language integration. For example, an activity in python that returns a json serialized object called from a Ruby workflow. Is the workaround to manually create the ruby object from json?

Yes, that will be a Ruby hash (just like it would be a JS object). If you want to convert it to a Ruby class you would have to do that yourself. We can consider more advanced JSON-to-class conversion mechanisms later, but by default Ruby doesn't really have any without the "addition" approach which is not very cross-language compatible.